src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,156 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final String iOutOfBounds = "Index is out of bounds";
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    public MyLinkedList() {
+    }
+
     @Override
     public void add(T value) {
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException(iOutOfBounds);
+        }
+
+        if (index == 0) {
+            head = new Node<>(null, value, head);
+            if (size == 0) {
+                tail = head;
+            }
+        } else if (index == size) {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+        } else {
+            Node<T> previous = getNode(index - 1);
+            Node<T> newNode = new Node<>(previous, value, previous.next);
+            previous.next = newNode;
+            if (newNode.next != null) {
+                newNode.next.prev = newNode;
+            }
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(iOutOfBounds);
+        }
+
+        Node<T> current;
+
+        if (index < size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
```

Comment:
this logic should be contained in getNode() method and this method should be reused here instead of full search